### PR TITLE
Add crowds to optimizer cost function

### PR DESCRIPTION
**Pull request by markdewing**
## Proposed changes
Additional parallelism in the optimizer cost function for processing the samples by handling them with multiple crowds.

Input parameters are 'opt_num_crowds' and 'opt_crowd_size'. Each is set to one by default for now.   We'll need to look at what should be better defaults.
    
The bulk of evaluation done by checkConfigurations and correlatedSampling has been pulled into a lambda.
    
The temporary data structures and object lists have been extracted to a separate class OptimizerEvalData.
    
The RecordsOnNode, DerivRecords, and HDerivRecords members variables have had an underscore appended, to differentiate them from the lambda parameter names.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

laptop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Review by ye-luo — CHANGES_REQUESTED** [review]
Needed changes are minor.

File: src/type_traits/template_types.hpp [hunk 1]
```diff
@@ -59,6 +59,21 @@ static RefVector<T> convertUPtrToRefVector(const UPtrVector<T>& ptr_list)
...
 
+template<class T>
+static RefVector<T> convertUPtrToRefVectorSubset(const UPtrVector<T>& ptr_list, int offset, int len)
+{
+  // check upper bounds
+  assert(offset + len <= ptr_list.size());
+
+  RefVector<T> ref_list;
+  ref_list.reserve(len);
+  for (int i = offset; i < offset + len; i++)
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/type_traits/template_types.hpp:70]: It was already there above.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp [hunk 2]
```diff
@@ -21,18 +21,73 @@
...
+    h0_ptr_list_[ib].reset(H_KE_Node.makeClone(*pCopy, *psiCopy));
+
+    rng_ptr_list_[ib] = std::make_unique<RandomGenerator_t>(Rng);
+    h_ptr_list_[ib]->setRandomGenerator(rng_ptr_list_[ib].get());
+
+    rng_save_ptr_ = std::make_unique<RandomGenerator_t>(Rng);
+  }
+}
+
+void OptimizerEvalData::set_current_crowd_size(int curr_crowd_size)
```
Comment (PDoakORNL) [inline src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:64]: see later I don't think this is necessary and the semantics are kind of weird.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
It looks like the crowd evaluation is basically done except for the compute_all_from_scratch section.

I'd like to see the curr_crowd_size_ complication stripped out. See comments, I think it would be easy.

File: src/QMCDrivers/WFOpt/CostFunctionCrowdData.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  UPtrVector<TrialWaveFunction> wf_ptr_list_;
+  UPtrVector<ParticleSet> p_ptr_list_;
+  UPtrVector<QMCHamiltonian> h_ptr_list_;
+  UPtrVector<QMCHamiltonian> h0_ptr_list_;
+  UPtrVector<RandomGenerator_t> rng_ptr_list_;
+
+  // Saved RNG state to reset to before correlated sampling
+  std::unique_ptr<RandomGenerator_t> rng_save_ptr_;
+
+  RefVector<TrialWaveFunction> wf_list_;
```
Comment (PDoakORNL) [inline src/QMCDrivers/WFOpt/CostFunctionCrowdData.h:79]: Unless I'm missing something these are uninitialized and unused now.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]

File: src/QMCDrivers/WFOpt/CostFunctionCrowdData.h (hunk 3 above)
Comment (markdewing) [reply src/QMCDrivers/WFOpt/CostFunctionCrowdData.h]: Good point.

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
@PDoakORNL OK with recent updates?

**Review by PDoakORNL — APPROVED** [review]
It would be nice to bring the naming in these new files in line with the coding standards.  Variable names seem a particular issue. I'm fine with this being addressed in a following PR.

**Pull request merged**